File: c/examples/dummy.c
/*
 * dummy.c — v0.10.3 OCXcip_Dummy liveness probe validator.
 *
 * Calls bp_client_dummy N times (default 100), measures the average
 * round-trip latency, and asserts rc == BP_OK for every call.
 *
 * SPDX-License-Identifier: MIT
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "bpclient.h"

static double now_us(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec * 1e6 + (double)ts.tv_nsec / 1e3;
}

int main(int argc, char *argv[]) {
    int n = 100;
    if (argc >= 2) n = atoi(argv[1]);
    if (n <= 0) n = 100;

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "[dummy] open failed\n");
        return 2;
    }

    int fail = 0;
    double t0 = now_us();
    for (int i = 0; i < n; i++) {
        int rc = bp_client_dummy(cl);
        if (rc != BP_OK) {
            fprintf(stderr, "[dummy] call %d rc=%d (%s)\n", i, rc, bp_strerror(rc));
            fail++;
        }
    }
    double dt = now_us() - t0;
    printf("[dummy] %d calls, %.0f us total, %.1f us/call, %d failures\n",
           n, dt, dt / (double)n, fail);
    printf("[dummy] %s\n", fail == 0 ? "PASS" : "FAIL");
    bp_client_close(cl);
    return fail == 0 ? 0 : 1;
}

File: python/examples/dummy.py
#!/usr/bin/env python3
"""dummy — v0.10.3 OCXcip_Dummy liveness-probe validator.

Mirrors c/examples/dummy.c + go/cmd/dummy.

SPDX-License-Identifier: MIT
"""
import sys
import time

import bpclient


def main() -> int:
    n = 100
    if len(sys.argv) >= 2:
        try:
            n = max(1, int(sys.argv[1]))
        except ValueError:
            pass

    try:
        c = bpclient.Client()
        c.open()
    except Exception as e:
        print(f"[dummy] open failed: {e}", file=sys.stderr)
        return 2

    fail = 0
    t0 = time.monotonic()
    try:
        for i in range(n):
            try:
                c.dummy()
            except Exception as e:
                print(f"[dummy] call {i} err={e}", file=sys.stderr)
                fail += 1
    finally:
        c.close()
    dt_us = (time.monotonic() - t0) * 1e6
    print(f"[dummy] {n} calls, {dt_us:.0f} us total, "
          f"{dt_us / n:.1f} us/call, {fail} failures")
    print(f"[dummy] {'PASS' if fail == 0 else 'FAIL'}")
    return 0 if fail == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
